File: src/vibe_kanban_clone/services/projects.py
"""Project service."""

import re
import unicodedata

import structlog
from sqlalchemy import select
from sqlalchemy.exc import IntegrityError
from sqlalchemy.ext.asyncio import AsyncSession

from vibe_kanban_clone.models.project import Project
from vibe_kanban_clone.schemas.project import ProjectCreate, ProjectUpdate

logger = structlog.get_logger()


def slugify(name: str) -> str:
    """Convert a name to a URL-friendly slug."""
    s = unicodedata.normalize("NFKD", name).encode("ascii", "ignore").decode()
    s = re.sub(r"[^a-z0-9]+", "-", s.lower()).strip("-")
    return s or "project"


def _is_slug_integrity_error(exc: IntegrityError) -> bool:
    msg = str(exc.orig) if exc.orig else str(exc)
    return "projects.slug" in msg


async def _ensure_unique_slug(
    session: AsyncSession, base_slug: str, exclude_id: int | None = None
) -> str:
    """Append a numeric suffix until the slug is unique (best-effort pre-check)."""
    slug = base_slug
    counter = 2
    while True:
        stmt = select(Project).where(Project.slug == slug)
        if exclude_id is not None:
            stmt = stmt.where(Project.id != exclude_id)
        result = await session.execute(stmt)
        if result.scalar_one_or_none() is None:
            return slug
        slug = f"{base_slug}{counter}"
        counter += 1


async def create_project(session: AsyncSession, data: ProjectCreate) -> Project:
    """Create a new project with a deduplicated slug."""
    base_slug = slugify(data.name)
    slug = await _ensure_unique_slug(session, base_slug)
    project = Project(
        name=data.name,
        slug=slug,
        description=data.description,
    )
    session.add(project)
    try:
        await session.commit()
    except IntegrityError as exc:
        await session.rollback()
        if not _is_slug_integrity_error(exc):
            raise
        counter = 2
        while True:
            project.slug = f"{base_slug}{counter}"
            counter += 1
            session.add(project)
            try:
                await session.commit()
                break
            except IntegrityError as inner_exc:
                await session.rollback()
                if not _is_slug_integrity_error(inner_exc):
                    raise
    await session.refresh(project)
    logger.info("project_created", project_id=project.id, slug=project.slug)
    return project


async def get_project(session: AsyncSession, project_id: int) -> Project | None:
    """Get a project by ID."""
    result = await session.execute(select(Project).where(Project.id == project_id))
    return result.scalar_one_or_none()


async def list_projects(session: AsyncSession, limit: int = 50, offset: int = 0) -> list[Project]:
    """List all projects."""
    result = await session.execute(select(Project).limit(limit).offset(offset))
    return list(result.scalars().all())


async def update_project(session: AsyncSession, project: Project, data: ProjectUpdate) -> Project:
    """Update a project, regenerating the slug if the name changes."""
    base_slug = None
    if data.name is not None:
        project.name = data.name
        base_slug = slugify(data.name)
        project.slug = await _ensure_unique_slug(session, base_slug, exclude_id=project.id)
    if data.description is not None:
        project.description = data.description

    if base_slug is not None:
        try:
            await session.commit()
        except IntegrityError as exc:
            await session.rollback()
            if not _is_slug_integrity_error(exc):
                raise
            counter = 2
            while True:
                project.name = data.name
                project.slug = f"{base_slug}{counter}"
                counter += 1
                if data.description is not None:
                    project.description = data.description
                try:
                    await session.commit()
                    break
                except IntegrityError as inner_exc:
                    await session.rollback()
                    if not _is_slug_integrity_error(inner_exc):
                        raise
    else:
        await session.commit()

    await session.refresh(project)
    logger.info("project_updated", project_id=project.id)
    return project


async def delete_project(session: AsyncSession, project: Project) -> None:
    """Delete a project."""
    await session.delete(project)
    await session.commit()
    logger.info("project_deleted", project_id=project.id)

File: server/src/services/projects.ts
import { Prisma } from "@prisma/client";
import { prisma } from "../prisma/client.js";
import { logger } from "../lib/logger.js";
import { slugify } from "../lib/utils.js";
import type { ProjectCreate, ProjectUpdate } from "../schemas/project.js";

export async function createProject(data: ProjectCreate) {
  const baseSlug = slugify(data.name);
  let slug = baseSlug;
  let counter = 2;
  while (true) {
    const existing = await prisma.project.findUnique({ where: { slug } });
    if (!existing) break;
    slug = `${baseSlug}${counter}`;
    counter++;
  }

  const project = await prisma.project.create({
    data: { name: data.name, slug, description: data.description },
  });
  logger.info({ project_id: project.id, slug }, "project_created");
  return project;
}

export async function getProject(projectId: number) {
  return prisma.project.findUnique({ where: { id: projectId } });
}

export async function listProjects(limit = 50, offset = 0) {
  return prisma.project.findMany({ take: limit, skip: offset });
}

export async function updateProject(projectId: number, data: ProjectUpdate) {
  const project = await getProject(projectId);
  if (!project) return null;

  const updateData: Prisma.ProjectUpdateInput = {};
  if (data.name !== undefined) {
    updateData.name = data.name;
    const baseSlug = slugify(data.name);
    let slug = baseSlug;
    let counter = 2;
    while (true) {
      const existing = await prisma.project.findFirst({
        where: { slug, NOT: { id: projectId } },
      });
      if (!existing) break;
      slug = `${baseSlug}${counter}`;
      counter++;
    }
    updateData.slug = slug;
  }
  if (data.description !== undefined) {
    updateData.description = data.description;
  }

  const updated = await prisma.project.update({
    where: { id: projectId },
    data: updateData,
  });
  logger.info({ project_id: updated.id }, "project_updated");
  return updated;
}

export async function deleteProject(projectId: number) {
  await prisma.project.delete({ where: { id: projectId } });
  logger.info({ project_id: projectId }, "project_deleted");
}
